### ISSUE TITLE
Expose request's body send progress in HTTPClient

### ISSUE THREAD

**Issue by vicguedez** [godotengine/godot-proposals#1221]
**Describe the project you are working on:**
Administrative application for a company.

**Describe the problem or limitation you are having in your project:**
Currently, in HTTPClient there is no way to get current "body" upload progress.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
It allows us to get the current progress of "upload file" requests so the end-user knows whats happening.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Make HTTPClient emit a signal when progress has been made in the request, send along with the signal how many has been sent (not the total amount, only what has just been sent).

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Because it must modify a core class.

**Comment by jonbonazza** [comment]
I agree that this could be useful for streaming large http response bodies allowing for faster, more efficient processing under certain scenarios. 

Might also be good to signal when full chunks have been read during chunked transfer encoding types

**Comment by madebydavid** [comment]
Currently there is no way to track the progress of an HTTP upload in Godot - this would be really useful to have.

I don't think it would need a signal - as the **download** progress does not work like that: 
- on `HTTPClient` you need to  add up the size of the chunks after calling `read_response_body_chunk()` and divide by `get_response_body_length()`
- on `HTTPRequest` you call `get_downloaded_bytes()`  and divide by `get_body_size()` 

An equivalent for **upload** progress would be:
- on `HTTPClient` expose a `get_request_bytes_sent()`  - the caller can divide by the known size of the file being uploaded
- on `HTTPRequest` expose a `get_uploaded_bytes()` - again divide by the known size

To expose the number of bytes sent - i think that could come from `request_buffer->get_position()` inside `HTTPClientTCP`

A complication here is that there is an `HTTPClientWeb` too - used for the web platform - and under the hood this uses the browser `fetch` function. The interface would need to be the same - and to implement the progress tracking via `fetch` there are some browser compatibility issues - https://stackoverflow.com/a/69400632/2779152 - looks like it will not work in Firefox.

I would be happy to work on a PR for this if there is interest and if people think I am going in the right direction.

**Comment by treepumpkin** [comment]
> I would be happy to work on a PR for this if there is interest and if people think I am going in the right direction.

Would have been a great feature to show build upload progress in the Hathora Godot plugin. Not planning to update it anytime soon, but there are other similar plugins that could benefit 😄

**Comment by madebydavid** [comment]
I have made the change to my own [fork here](https://github.com/madebydavid/godot/commit/6da588328298e8982d68e243d88550cbcc605769) and have been doing some testing with an [example project](https://github.com/madebydavid/godot-http-upload-example). 

The approach we have discussed works, but there is something we should be aware of - `request_buffer->get_position()` resets to 0 as soon as we have finished sending to the server - but before the request is completed - this is because the internal request_buffer is cleared and we are still waiting for a response from the server.

This means that a simple progress calculation like this:

```gdscript
var headers := ["Content-Type: multipart/form-data; boundary=%s" % multipart.boundary]
http.request_raw(HTTPClient.METHOD_POST, parsed.path, headers, body)

while http.get_status() == HTTPClient.STATUS_REQUESTING:
	http.poll()
	var percent := (http.get_request_bytes_sent() * 100.0) / total_bytes
	log_message("Uploading: %.1f%%" % percent)
	await get_tree().process_frame

log_message("Response: %d" % http.get_response_code())
http.close()
```

Will have output that looks like this, where the %age seems to drop to zero once the buffer has fully sent:

```
...
[12:53:14] Uploading: 95.8%
[12:53:14] Uploading: 96.6%
[12:53:14] Uploading: 97.5%
[12:53:14] Uploading: 98.4%
[12:53:14] Uploading: 99.3%
[12:53:14] Uploading: 0.0%
[12:53:14] Uploading: 0.0%
[12:53:14] Uploading: 0.0%
[12:53:14] Uploading: 0.0%
[12:53:14] Response: 200
...
```

Checking if `get_request_bytes_sent()` > 0 solves this - but it will likely be a gotcha for some users:

```gdscript
while http.get_status() == HTTPClient.STATUS_REQUESTING:
    http.poll()
    var sent := http.get_request_bytes_sent()
    if sent > 0:
        var percent := (sent * 100.0) / total_bytes
        log_message("Uploading: %.1f%%" % percent)
    await get_tree().process_frame

```

This could be solved in the HTTPClient by persisting the `request_buffer->get_position()` in a local variable until the status changes from `HTTPClient.STATUS_REQUESTING`.

Should I persist the counter value until status changes, or do you think is this acceptable? This would be my first PR for godot.
